sem/utils.py

```diff
@@ -481,3 +490,327 @@ def convert_environment_str_to_dict(log_components):
+
+    Args:
+        db (TinyDB instance): A TinyDB instace where the logs are inserted.
+        data_dir (str): Path to where the TinyDB instance is stored
+    """
+    # Clean logs table
+    db.drop_table('logs')
+    db.storage.flush()
+
+    # Get rid of contents of data dir
+    try:
+        os.remove(data_dir)
+    except OSError as error:
+        print(error)
+        print("File path '%s' can not be removed" % data_dir)
+
+
+def filter_logs(db,
+                severity_class=None,
+                components=None,
+                function=None,
+                context=None,
+                time_begin=None,
+                time_end=None):
+    """
+    Filter the logs stored in the database.
+
+    Filters are applied on context, function name, log sevirity class and time.
+    Additionally the user can also filter each log component based on a
+    particular sevirity class using components dictionary.
+    For example, if the user specifies Context = [0, 1] and Function = [A, B]
+    the function will output logs in which (context == 0 or context == 1) and
+    (function == a or function == b).
+
+    Return a list of logs that satisfy all the passed filters. Each log is
+    represented by a dictionary.
+
+    Args:
+        db (TinyDB instance): A TinyDB instace where the logs are inserted.
+        context (list, str, int): A list of context based on which the logs
+            will be filtered. If only one context is to be provided, then this
+            paramter can also be a string or an int.
+        function (list, str): A list of function names based on which the logs
+            will be filtered. If only one function name is to be provided, then
+            this paramter can also be a string.
+        time_begin (float): Start timestamp (in seconds) of the time window.
+        time_end (float): End timestamp (in seconds) of the time window.
+        severity_class (list, str): A list of log severity classes based on
+            which the logs will be filtered. If only one log severity class is
+            to be provided, then this paramter can also be a string.
+        components (dict): A dictionary having structure
+            {
+                components:['class1','class2']
+            }
+            based on which the logs will be filtered.
+    """
+    query_final = []
+
+    # Assert that the passed paramters are of valid type.
+    if severity_class is not None:
+        if severity_class is not None:
+            if isinstance(severity_class, str):
+                severity_class = [severity_class]
+            elif isinstance(severity_class, list):
+                pass
+            else:
+                raise TypeError("severity_class can only be a list or a string (if only one value is passed).")
+
+    if components is not None:
+        for key, value in components.items():
+            if isinstance(value, str):
+                components[key] = [value]
```

Comment:
As discussed during the call, it might be better to make a copy of components before modifying it, so that we don't touch the dictionary outside of the function. The same applies to the other arguments, too.